### PR TITLE
Classic mode generates .cmx

### PR DESCRIPTION
**Pull request by Keryan-dev**
From #203, [relevant diff](https://github.com/ocaml-flambda/flambda-backend/pull/336/files/596d716850cd2e020ecd7079b587eaf0cfbf4d5e..6ae749cc3f43856317708e825956c2a9f9e6c46c).

This improves #203 by generating .cmx for units compiled in classic mode, allowing the inlining to be done beyond the local unit.

Thanks to the work of @lthls, the cmx of both classic mode and simplify are compatible.

### CODE REVIEW

File: middle_end/flambda2/simplify/inlining/inlining_report.mli
Comment (Gbury) [inline middle_end/flambda2/simplify/inlining/inlining_report.mli:36]: you need to add a `(**)` before the actual doc comment, else the doc comment is attached to the last constructor and not the whole type declaration

**Review by Gbury — DISMISSED** [review]
Looks good overall (though I didn't look at the cmx handling changes). However, this PR conflicts with other re-factoring PR currently opened, and I suspect the rebase will be non trivial, so let's wait until the re-factoring PR are merged before merging this one ?

File: middle_end/flambda2/simplify/inlining/inlining_report.mli
Comment (lukemaurer) [reply middle_end/flambda2/simplify/inlining/inlining_report.mli:36]: Looks like there's already a doc comment for the type declaration. Probably an oversight?

File: middle_end/flambda2/from_lambda/closure_conversion_aux.ml
Comment (lthls) [inline middle_end/flambda2/from_lambda/closure_conversion_aux.ml:424]: You should replace this comment with a note explaining why the continuation is kept in `continuation_applications`

**Review by lthls — APPROVED** [review]
Looks good to me, but I haven't reviewed my own contributions too much

**Review by poechsel — APPROVED** [review]

File: middle_end/flambda2/types/flambda2_types.ml
Comment (lthls) [inline middle_end/flambda2/types/flambda2_types.ml:92]: The answer is no, it can happen if the function isn't inlinable for instance. in this case it's still useful to have the approximation, to be able to generate direct calls. I'll remove the comment.

File: middle_end/flambda2/types/flambda2_types.ml
Comment (lthls) [inline middle_end/flambda2/types/flambda2_types.ml:117]: This is a bit fishy, but I'm not sure `Local` allocations reachable from a symbol make sense anyway

File: middle_end/flambda2/types/env/typing_env.ml
Comment (lthls) [inline middle_end/flambda2/types/env/typing_env.ml:1266]: This should be `[]`, I think. All symbols are added to the `names_to_types` map, so they all have equations in the result. I'll check that it doesn't obviously break anything, and push the change.

File: middle_end/flambda2/types/env/typing_env.ml
Comment (lthls) [inline middle_end/flambda2/types/env/typing_env.ml:1281]: Note: this horrible hack is going to cause problems later (the closure ID could show up in the free names of the typing env later, leading to errors during the offset computations). In this PR we don't use the free names of the generated typing env, and @Keryan-dev has a cleaner fix that keeps the closure ID in the approximation as part as another PR, so we're leaving it as it is for now.

**Review by lthls — COMMENTED** [review]
I've re-reviewed the PR. It's mostly fine, but I've left comments in a few places that we'll need to come back to at some point.

There are also a few small patches I'd like to make to the conversions between types and approximations, I'll approve when I've pushed them.

**Review by lthls — APPROVED** [review]

File: middle_end/flambda2/types/env/typing_env.ml
Comment (lthls) [reply middle_end/flambda2/types/env/typing_env.ml]: Done.

File: middle_end/flambda2/types/flambda2_types.ml
Comment (lthls) [reply middle_end/flambda2/types/flambda2_types.ml]: Done.

File: middle_end/flambda2/cmx/flambda_cmx.ml
Comment (lthls) [inline middle_end/flambda2/cmx/flambda_cmx.ml:247]: We could add a `|| not (Flambda_features.classic_mode ())` to save some time when not in classic mode.

File: middle_end/flambda2/cmx/flambda_cmx.ml
Comment (Keryan-dev) [reply middle_end/flambda2/cmx/flambda_cmx.ml:252]: Addressed in #571

File: middle_end/flambda2/from_lambda/closure_conversion.ml
Comment (Keryan-dev) [reply middle_end/flambda2/from_lambda/closure_conversion.ml:1706]: Addressed in #571

File: middle_end/flambda2/types/env/typing_env.ml
Comment (Keryan-dev) [reply middle_end/flambda2/types/env/typing_env.ml:1281]: Addressed in #424

File: middle_end/flambda2/from_lambda/closure_conversion_aux.ml
Comment (lthls) [reply middle_end/flambda2/from_lambda/closure_conversion_aux.ml:160]: We do filter on export now, but the function isn't completely redundant as we could be loading a cmx compiled with Simplify (which has slightly different filtering heuristics). So we're going to leave it as is for now.

**Pull request merged**
